blog/_src/posts/2016-07-25-tutorial-zero-to-sixty-in-racket.scrbl

```diff
@@ -0,0 +1,698 @@
+
+When called with a filename like @racket{heart-of-darkness.txt}, the function
+ uses @hyperlink["http://docs.racket-lang.org/reference/for.html#%28form._%28%28lib._racket%2Fprivate%2Fbase..rkt%29._for%2Flist%29%29"]{for/list} to build a list of lines by reading from a port
+ with @hyperlink["http://docs.racket-lang.org/reference/sequences.html#%28def._%28%28lib._racket%2Fprivate%2Fbase..rkt%29._in-lines%29%29"]{in-lines}.
+The port is the data from @racket[filename], thanks to @hyperlink["http://docs.racket-lang.org/reference/file-ports.html#%28def._%28%28lib._racket%2Fprivate%2Fbase..rkt%29._with-input-from-file%29%29"]{with-input-from-file}.
+
+Next is a function to divide a list of strings into a list of lists of words.
+Here we'll just use library functions.
+
+@(begin
+  #reader scribble/comment-reader
+  (racketblock
+    (define (kwik-split lines)
+      (map string-split lines))
+))
+
+By default, @hyperlink["http://docs.racket-lang.org/reference/strings.html#%28def._%28%28lib._racket%2Fstring..rkt%29._string-split%29%29"]{string-split} divides a string into a list of whitespace-separated substrings.
+You can always supply a different delimiter, or use @hyperlink["http://docs.racket-lang.org/reference/regexp.html#%28def._%28%28lib._racket%2Fprivate%2Fbase..rkt%29._regexp-split%29%29"]{regexp-split}
+ to divide by a regular expression.
+
+Two tasks left!
+First we generate all circular shifts for a list of strings @racket[words]
+ by folding up a list with one shift of @racket[words] for each word.
+
+@(begin
+  #reader scribble/comment-reader
+  (racketblock
+
+    (define (circular-shift words)
+      (append (rest words) (list (first words))))
+
+    (define (all-circular-shifts words)
+      (for/fold ([all-shifts (list words)])
+                ([i (in-range 1 (length words))])
+        (cons (circular-shift (first all-shifts))
+              all-shifts)))
+))
+
+Second, we alphabetize and print the shifts.
+
+@(begin
+ #reader scribble/comment-reader
+ (racketblock
+  (define (alphabetize all-shifts)
+    (sort all-shifts shift<?))
+
+  (define (shift<? shift1 shift2)
+    (match* (shift1 shift2) ; destruct multiple values
+     [('() _) ; first list empty, don't care about second
+      #t]
+     [(_ '()) ; first list non-empty, second empty
+      #f]
+     [((cons s1 shift1-rest) (cons s2 shift2-rest))
+      (or (string<? s1 s2)
+          (and (string=? s1 s2)
+               (shift<? shift1-rest shift2-rest)))]))
+
+  (define (kwik-display all-sorted-shifts)
+    (define (display-words words)
+      (display (first words))
+      (for ([word (in-list (cdr words))])
+        (display " ")
+        (display word))
+      (newline))
+    ; for-each is like map, but always returns (void)
+    (for-each display-words all-sorted-shifts))
+))
+
+Gluing it all together, here's the full script.
+
+@codeblock{
+  #lang racket
+
+  ; type Words = (Listof String)
+  ; type Lines = (Listof Words)
+
+  ; Path-String -> (Listof String)
```

Comment:
Wait, I'm not sure.

First thought: "wow how could I forget contracts, they absolutely belong here"

Second thought: "I'd have to write them as `define/contract`." But that's not how contracts are actually used. 

Should I just use `define/contract`? Is there another way to get the point across in this single-module world?